File: be_pay/be_pay/events/leave_application.py
import frappe

def before_submit(doc, method=None):
    # --- from script: Leave Off validate ---
    start_date = doc.from_date
    end_date = doc.to_date

    # Convertir les dates en objets datetime
    start_date_obj = frappe.utils.getdate(start_date)
    end_date_obj = frappe.utils.getdate(end_date)

    # Calculer la différence en jours entre les deux dates
    time_diff = end_date_obj - start_date_obj
    total_days = time_diff.days + 1  # +1 pour inclure le jour de départ

    # Vérifier chaque date dans la table custom_jours_off
    valid_days_off = 0
    for row in doc.custom_jours_off :  # Parcourir la liste en sens inverse

        row_date = frappe.utils.getdate(row.date_day_off)

        valid_days_off = valid_days_off + 1  # Compter les jours valides dans la période

    # Calculer le nombre de jours de congé en soustrayant les jours valides
    leave_days = total_days - valid_days_off

    # Afficher le nombre de jours de congé calculé
    #frappe.msgprint(_(f"Nombre de jours de congé calculé : {leave_days}"))
    doc.total_leave_days = leave_days



def after_save(doc, method=None):
    # --- from script: Leave Swit Off ---
    start_date = doc.from_date
    end_date = doc.to_date

    # Convertir les dates en objets datetime
    start_date_obj = frappe.utils.getdate(start_date)
    end_date_obj = frappe.utils.getdate(end_date)

    # Calculer la différence en jours entre les deux dates
    time_diff = end_date_obj - start_date_obj
    total_days = time_diff.days + 1  # +1 pour inclure le jour de départ

    # Vérifier chaque date dans la table custom_jours_off
    valid_days_off = 0
    for row in doc.custom_jours_off :  # Parcourir la liste en sens inverse

        row_date = frappe.utils.getdate(row.date_day_off)

        valid_days_off = valid_days_off + 1  # Compter les jours valides dans la période

    # Calculer le nombre de jours de congé en soustrayant les jours valides
    leave_days = total_days - valid_days_off

    # Afficher le nombre de jours de congé calculé
    #frappe.msgprint(_(f"Nombre de jours de congé calculé : {leave_days}"))
    doc.total_leave_days = leave_days



def after_submit(doc, method=None):
    # --- from script: Add Attendance ---
    from_date = frappe.utils.getdate(doc.from_date)
    to_date = frappe.utils.getdate(doc.to_date)
    current_date = from_date

    # Créer un tableau pour stocker les enregistrements d'assiduité
    attendance_records = []

    while current_date <= to_date:
        # Vérifier s'il y a un jour férié ou un jour de congé
        query_holiday = """
            SELECT * FROM `tabHoliday` WHERE holiday_date = %s AND weekly_off = 1
        """
        holiday_results = frappe.db.sql(query_holiday, (current_date,), as_dict=True)

        # Vérifier si un enregistrement de présence existe déjà pour cette date
        query_attendance = """
            SELECT name FROM `tabAttendance` 
            WHERE employee = %s AND attendance_date = %s AND shift = %s
        """
        attendance_exists = frappe.db.sql(query_attendance, 
                                          (doc.employee, current_date, 'O' if holiday_results else doc.leave_type), 
                                          as_dict=True)

        if attendance_exists:
            frappe.msgprint(f"L'assiduité pour l'employé {doc.employee} est déjà marquée pour la date {current_date}.")
        else:
            # Créer une instance de Attendance List
            attendance_list = frappe.new_doc('Attendance')

            # Remplir les champs principaux
            attendance_list.employee = doc.employee
            attendance_list.employee_name = doc.employee_name
            attendance_list.status = "On Leave"

            #if holiday_results:
               # attendance_list.leave_type = 'O'
                #attendance_list.shift = 'O'
            #else:
            attendance_list.leave_type = doc.leave_type
            attendance_list.shift = doc.leave_type

            attendance_list.leave_application = doc.name
            attendance_list.attendance_date = current_date
            attendance_list.company = doc.company
            attendance_list.department = doc.department

            # Ajouter l'enregistrement à la liste
            attendance_records.append(attendance_list)

        # Passer au jour suivant
        current_date = frappe.utils.add_days(current_date, 1)

    # Sauvegarder et soumettre tous les enregistrements
    for attendance in attendance_records:
        attendance.save()
        attendance.submit()

    # Message de confirmation
    frappe.msgprint("Tous les enregistrements de présence ont été créés avec succès.")

File: be_pay/public/js/leave_application.js
// --- Script for Leave Application ---
frappe.ui.form.on('Leave Application', {
	employee: function(frm) {
		frm.trigger("make_dashboard2");
	},
	from_date: function(frm) {
		frm.trigger("make_dashboard2");
	},
	to_date: function(frm) {
		frm.trigger("make_dashboard2");
	},
	
	make_dashboard2: function(frm) {
		var leave_details;
		let all_leaves;
		if (frm.doc.employee) {
			frm.set_query('leave_type', function() {
				return {
					filters: [
						['leave_type_name', '!=', 'zzz']
					]
				};
			});
		}
	},
})

// --- Script for Leave Application ---
frappe.ui.form.on('Leave Application', {
  refresh(frm) {
  // Votre code ici (si nécessaire)
  },
 
  before_save: function(frm) {
  let start_date = frm.doc.from_date;
  let end_date = frm.doc.to_date;
  
  let start_date_obj = new Date(start_date);
  let end_date_obj = new Date(end_date);
  
  // Calculer la différence en jours entre les deux dates
  let timeDiff = end_date_obj - start_date_obj;
  let totalDays = Math.ceil(timeDiff / (1000 * 60 * 60 * 24)) + 1; // +1 pour inclure le jour de départ
 
  // Vérifier chaque date dans la table custom_jours_off, si elle existe
  let validDaysOff = 0;
  if (frm.doc.custom_jours_off && frm.doc.custom_jours_off.length > 0) { // Vérifier si le tableau n'est pas vide
  for (let i = frm.doc.custom_jours_off.length - 1; i >= 0; i--) {
  let row = frm.doc.custom_jours_off[i];
  let row_date = new Date(row.date_day_off);
  
  if (row_date < start_date_obj || row_date > end_date_obj) {
  frm.doc.custom_jours_off.splice(i, 1);
  frappe.msgprint(__("La date " + row.date_day_off + " ne se trouve pas dans la période délimitée. Veuillez choisir une autre date."));
  } else {
  validDaysOff++; // Compter les jours valides dans la période
  }
  } 
  }
 
  // Calculer le nombre de jours de congé en soustrayant les jours valides
  let leaveDays = totalDays - validDaysOff;
 
  // Afficher le nombre de jours de congé calculé
  // frappe.msgprint(__("Nombre de jours de congé calculé : " + leaveDays));
  frm.set_value('total_leave_days', leaveDays);
 
  // Rafraîchir le champ pour refléter les modifications
  frm.refresh_field('custom_jours_off');
  }
 });
 

